### DISCUSSION TITLE
How to extend mdast?

### DISCUSSION THREAD

**Discussion by eliyap** [micromark/micromark#83] [Q&A] [answered]
Hi All!

I wanted to implement Pandoc's ==highlight== [syntax](https://stackoverflow.com/questions/25104738/text-highlight-in-markdown), and since the existing [GFM strikethrough](https://github.com/syntax-tree/mdast-util-gfm-strikethrough) is very similar, I thought I would start there.

However, I noticed that the type annotations rely on a pre-existing "Delete" extension. What is the correct way to extend PhrasingContent to include, say, a new "Mark" type (to correspond to the html <mark> tag)?

**Comment by wooorm** [comment, marked as answer]
Hey!

One part of how to extend mdast is conceptual, and you already found it. Your mark syntax is **Phrasing** content.
The other part is TypeScript-specific. You can define a new `Mark` node that extends from `Parent`, with a `type: 'mark'`, and `children: PhrasingContent[]`, and then register the mark node by adding it to `StaticPhrasingContentMap`:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fe99cda97743cac0940e4960e3a904b655277362/types/mdast/index.d.ts#L76-L101

**Reply by wooorm** [reply]
Here is an example of how directives are added: https://github.com/syntax-tree/mdast-util-directive/blob/778e22e790ed85dcc9f04bcd8ac02e33a55d39e7/complex-types.d.ts#L26-L35.
Note that we typically have lower-level packages such as `mdast-util-directive`, which are then wrapped in a remark package: `remark-directive`. We define stuff in the lower-level packages and then import them, such as like this: https://github.com/remarkjs/remark-directive/blob/2631e46657a401e32c7555342d30b58bc57d6d4f/index.js#L4, to include the registered nodes for anyone that uses `remark-directive`.

**Reply by eliyap** [reply]
I'm still having trouble navigating this extension, not familiar with typing / JS. A few architecture questions:
1. are `.d.ts` files meant to be _written_ or _generated_? If generated, are the declarations written in JSDoc? 
2. Should source files be written in `.js` or `.ts`? I noticed the [GFM Strikethrough](https://github.com/syntax-tree/mdast-util-gfm-strikethrough) has only `.js` source files(?)

**Reply by ChristianMurphy** [reply]
> written or generated?

Either is fine, if you are using `.ts` files or `.js` with type checked JSDoc, generated may be easier.
If you are using plain `.js` with no type checking, hand writing the `.d.ts` may be easier

> If generated, are the declarations written in JSDoc?

`declare module` can be written directly in a `.ts` file or a `.d.ts`, I don't think TS supports module extension from `.js` with JSDoc just yet (though it would be nice to have)

> Should source files be written in `.js` or `.ts`?

Which do you prefer? That's the one to pick.
If you don't have a preference, `.ts` may be easier to get started with, there are more resources/documentation than JSDoc based typings.
Then if you want to skip the compile step and use plain `.js`, JSDoc based TypeScript is still an option that you could move to down the road.

**Reply by ChristianMurphy** [reply]
Some other resources which can help, the TypeScript handbook is full of helpful information and pointers https://www.typescriptlang.org/docs/handbook/intro.html, and TypeScript deep dive is also good https://basarat.gitbook.io/typescript/

When generating typings, I'd recommend setting `strict` mode (https://www.typescriptlang.org/tsconfig#strict) and checking the code with `type-coverage` to ensure most variables are typed (https://github.com/plantain-00/type-coverage).

If hand writing types, the TypeScript handbook has a good intro (https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html).
I'd recommend using dtslint (https://github.com/Microsoft/dtslint) to check the typings, and following the best practices from definitely typed https://definitelytyped.org/guides.html as well.

**Reply by eliyap** [reply]
[So far so good](https://github.com/eliyap/mdast-highlight)! The interface extensions and `d.ts` generation works great, thank you for the pointers!

I did run into a sharp corner with the whole ES6/RequireJS/CommonJS mess, specifically this error:

```
$ tsc && node --conditions development test.js

/Users/SAM/Documents/mdast-highlight/mdast-highlight/test.js:4
var mdast_util_from_markdown_1 = require("mdast-util-from-markdown");
                                 ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/SAM/Documents/mdast-highlight/mdast-highlight/node_modules/mdast-util-from-markdown/dev/index.js from /Users/SAM/Documents/mdast-highlight/mdast-highlight/test.js not supported.
Instead change the require of index.js in /Users/SAM/Documents/mdast-highlight/mdast-highlight/test.js to a dynamic import() which is available in all CommonJS modules.
    at Object.<anonymous> (/Users/SAM/Documents/mdast-highlight/mdast-highlight/test.js:4:34) {
  code: 'ERR_REQUIRE_ESM'
}
```

The error seems to be caused by `var mdast_util_from_markdown_1 = require("mdast-util-from-markdown");`, and replacing that with `var mdast_util_from_markdown_1 = import("mdast-util-from-markdown");` fixes it. I understand that UNIST has gone ESM only, so that makes sense.

Is there a way to make TSC not use `require` in the exported `.js`? I tried building and running the [gfm strikethrough `test.js`](https://github.com/syntax-tree/mdast-util-gfm-strikethrough/blob/main/test.js) on my machine to see what the difference was, but it encountered a similar error.

**Reply by eliyap** [reply]
My Bad 😞
Apparently I needed to target ES2020, which I hadn't noticed in the tsconfig. All is good!

**Reply by eliyap** [reply]
Hello again!

Here's [micromark-extension-pandoc-highlight](https://github.com/eliyap/micromark-extension-pandoc-highlight), now on [npm](https://www.npmjs.com/package/micromark-extension-pandoc-highlight). It passes a tweaked version of the strikethrough tests.

Here's [mdast-highlight](https://github.com/eliyap/mdast-highlight), which imports the above package. It's half working: the parse fails, but the export parses:

Here's a truncated test log:
```
# markdown -> mdast
not ok 1 should support highlight
expected: |-
      { type: 'root', children: [ { type: 'paragraph', children: [ { type: 'text', value: 'a ' }, { type: 'mark', children: [Array] }, { type: 'text', value: ' c.' } ] } ] }
    actual: |-
      { type: 'root', children: [ { type: 'paragraph', children: [ { type: 'text', value: 'a b c.' } ] } ] }

not ok 2 should support highlight w/ eols
expected: |-
      { type: 'root', children: [ { type: 'paragraph', children: [ { type: 'text', value: 'a ' }, { type: 'mark', children: [Array] }, { type: 'text', value: ' d.' } ] } ] }
    actual: |-
      { type: 'root', children: [ { type: 'paragraph', children: [ { type: 'text', value: 'a ==b\nc== d.' } ] } ] }
# mdast -> markdown
ok 3 should serialize highlight
ok 4 should serialize highlight w/ eols
```

My question is: why is `fromMarkdown` missing the `mark`? 

I suspect it's because I was wishy washy about "highlight" vs "mark".
But `gfm-strikethrough` is able to connect `delete` to `strikethrough` somewhere, I'm just not sure where that connection is. Any comments about code quality / suggestions also welcome!

**Reply by wooorm** [reply]
#### Name

- A name like `mdast-util-pandoc-highlight` might be better for the utility, `util-` because all utilities have that prefix, and `pandoc-highlight` might be a more explanatory name in my opinion and is in line with your micromark extension?
- Alternatively the utility uses `pandocMark` as a name internally, I think the name exported by the extension should be `x` (where `x` is the camelcased form of the dash-cases `y` in `micromark-extension-y`), and then the utility should be called `mdast-util-y` and export `xFromMarkdown` and `xToMarkdown`
- Lastly, this seems to translate to an HTML `<mark>` element? Maybe just `micromark-extension-mark` and `mdast-util-mark`

#### Other

- I can’t see the code for the extension on GH?
- https://github.com/eliyap/mdast-highlight/blob/27f97d9a5fea04d867f5a3451def299c1aa6dc9a/index.js#L9 this should probably be `mark` instead of `emphasis`
- > I suspect it's because I was wishy washy about "highlight" vs "mark".

  I think your extension probably doesn’t use `mark` as a name, but your `fromMarkdown` extension does. Those should probably match up?

**Reply by eliyap** [reply]
> this should probably be `mark` instead of `emphasis`

I copied the code [from `gfm-strikethrough`](https://github.com/syntax-tree/mdast-util-gfm-strikethrough/blob/681c22b32b4da2797cba90c529bad16580d8ce0e/index.js#L41), should that be `delete` instead?

**Reply by wooorm** [reply]
Yes!

**Reply by eliyap** [reply]
> I can’t see the code for the extension on GH?

It was private 😬.
Renamed the repos to [micromark-extension-mark](https://github.com/eliyap/micromark-extension-mark) and [mdast-util-mark](https://github.com/eliyap/mdast-util-mark). 

Still working on finding the naming inconsistency.

**Reply by eliyap** [reply]
Fixed the inconsistency, it works!

Packages are on [npm](https://www.npmjs.com/settings/ipskie/packages). Thanks for all the help!

**Reply by shlroland** [reply]
@wooorm  @eliyap  Hi, I am working on implementing the same functionality, but I encountered this type of error.
<img width="1265" alt="image" src="https://github.com/micromark/micromark/assets/45024986/398f1b0e-e8ce-40e6-b732-9d32e3bfd426">
<img width="550" alt="image" src="https://github.com/micromark/micromark/assets/45024986/3d2762ec-f102-4e76-bcd4-da72cea5c2ac">

**Reply by ChristianMurphy** [reply]
@shlroland did you import the type definition that adds the `Mark` type in the file/project where you are attempting to add a `Mark` node?
Pulling from a related example in `remark-directive` the typing needs to be registered in the plugin where it is used https://github.com/remarkjs/remark-directive#types

**Reply by shlroland** [reply]
It's because my monorepo has multiple versions of TypeScript, causing this issue.

**Reply by ChristianMurphy** [reply]
Okay, so it sounds like you have a broader typing issue across your monorepo, and the typing you shared is just one of several ways it manifests?
I'm not sure how we can help outside of offering encouragement to upgrade your TypeScript versions to the latest 🙂

**Reply by shlroland** [reply]
When I switched back to the commit before upgrading to the latest TypeScript version, I found that the issue might not be related to TypeScript version inconsistency. The actual problem seems to be that when I upgraded the project's mdast to the latest version 4.0, some remark plugins were still dependent on the older version of mdast, which caused this issue to occur.

**Reply by ChristianMurphy** [reply]
@shlroland  `unified`, `micromark`, and most of packages on npm follow semantic versioning https://semver.org/
A major version means it is a breaking change and incompatible with previous versions.
`typescript` sometimes includes breaking changes in minor versions as well.
The version numbers and change logs are communicated so adopters (like yourself) have the information to choose versions which are compatible across projects.
If you choose incompatible versions, it won't work.

That fact doesn't particularly relate this this thread, which is about extending `mdast`.
If you have a specific question about extending the `mdast` types, feel free to ask it.
If you need general `micromark` support, feel free to start a new thread https://github.com/orgs/micromark/discussions/new?category=q-a
In either case, please review the support guidelines https://github.com/micromark/.github/blob/main/support.md for recommendations on how to ask questions so others can effectively help you.
